core/cont/inc/RangeCast.h

```diff
@@ -0,0 +1,185 @@
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RangeCast_h
+#define ROOT_RangeCast_h
+
+#include <cassert>
+#include <type_traits>
+#include <utility>
+
+namespace ROOT {
+namespace Internal {
+
+#if (__cplusplus < 201700L)
+
+template <typename T, bool isDynamic = false>
+struct Cast {
+   template <typename U>
+   static T cast(U &&u)
+   {
+      return dynamic_cast<T>(u);
+   }
+};
+
+template <typename T>
+struct Cast<T, false> {
+   template <typename U>
+   static T cast(U &&u)
+   {
+      assert(dynamic_cast<T>(u));
+      return static_cast<T>(u);
+   }
+};
+
+#endif
+
+// For SFINAE-based checks for the existence of the `begin` and `end` methods.
+template <typename T>
+constexpr auto hasBeginEnd(int) -> decltype(std::declval<T>().begin(), std::declval<T>().end(), true)
+{
+   return true;
+}
+
+template <typename>
+constexpr bool hasBeginEnd(...)
+{
+   return false;
+}
+
+} // namespace Internal
+} // namespace ROOT
+
+template <typename T, typename WrappedIterator_t, bool isDynamic>
+class TypedIter {
+
+public:
+   TypedIter(WrappedIterator_t const &iter) : _iter{iter} {}
+
+   TypedIter &operator++()
+   {
+      ++_iter;
+      return *this;
+   }
+   TypedIter operator++(int)
+   {
+      TypedIter tmp(*this);
+      operator++();
+      return tmp;
+   }
+   bool operator==(const TypedIter &rhs) const { return _iter == rhs._iter; }
+   bool operator!=(const TypedIter &rhs) const { return _iter != rhs._iter; }
+
+   void swap(TypedIter &other) { _iter.swap(other._iter); }
+
+   // We want to know at compile time whether dynamic_cast or static_cast is
+   // used. First of all to avoid overhead, but also to avoid a compiler
+   // error when using dynamic_cast on a non-polymorphic class. In C++17,
+   // this can be done easily with `if constexpr`, but for the older
+   // standards we have to use a more verbose alternative. Both ways are
+   // explicitely implemented for different standards, so that when the
+   // minimum C++ standard for ROOT is raised to C++17 it's easy to remember
+   // that we can avoid much boilerplate code in this file.
+#if (__cplusplus < 201700L)
+   T operator*() { return ROOT::Internal::Cast<T, isDynamic>::cast(*_iter); }
+#else
+   T operator*()
+   {
+      if constexpr (isDynamic) {
+         return dynamic_cast<T>(*_iter);
+      } else {
+         assert(dynamic_cast<T>(*_iter));
+         return static_cast<T>(*_iter);
+      }
+   }
+#endif
+   typename std::remove_reference<T>::type *operator->() { return &**this; }
```

Comment:
The problem here is that `T` *can be* a reference: it is the type you cast to, so it is usually a reference or a pointer type.
Usually it is, so the code will work.

But yes, if you cast to a different value like `static_cast<double>` this is problematic because of the pointer to a temporary like you say. As far as I can tell, we will always have this problem with `operator->` no matter which one of the two is implemented in terms of the other.

Maybe we should just not implement `operator->`? It is not necessary for the use of `RRageCast` is range-based loops. I'll remove the `operator->` in my next commit, let me know if you prefer something else.